content/post/051-new-ipfs-release-process.md

```diff
@@ -0,0 +1,229 @@
+---
+date: 2019-07-18
+url: 2019-07-18-ipfs-release-process
+tags: IPFS, go-ipfs, release process, 
+title: Improving the IPFS Release Process
+author: Steven Allen, Alan Shaw, David Dias, Molly Mackinlay
+---
+
+go-ipfs is introducing a new release cycle and process to ensure reliable releases.
+
+## Background
+
+When reflecting on Q2, we noticed an unusual number of critical regressions
+(since fixed) introduced in the last three go-ipfs releases:
+
+* **go-ipfs 0.4.21** had two performance regressions in bitswap:
+  1. A throughput regression that should have been caught by regression testing
+    (now tested) but almost certainly would have been noticed by downstream
+    users given a longer release-testing process.
+  2. A CPU utilization regression that only shows up with >10000 peers. Something
+    that only really shows up in certain production systems under heavy load.
+* **go-ipfs 0.4.20** had a regression where adding multiple independent files in the same
+  add command didn't work ([#6254](https://github.com/ipfs/go-ipfs/pull/6255)).
+  A regression test has since been added, but this _also_ would have been caught
+  with better cross-application testing.
+* **go-ipfs 0.4.19** had multiple regressions (all fixed in 0.4.20):
+  1. A regression in the docker container (introduced by
+    [#6040](https://github.com/ipfs/go-ipfs/pull/6040)) that would have been
+    caught by testing the go-ipfs docker image in more production environments.
+  2. A CPU utilization regression in bitswap only seen under very high load. This
+    would have been caught by testing under production loads.
+  3. Panics in the DHT and QUIC modules that only show up under heavy load.
+
+We found two root causes:
+
+1. An uptick in development pace with respect to previous quarters without
+   matching improvements to our testing practices. Several large refactors
+   touching key but poorly tested subsystems landed in Q2.
+2. A significant increase in the network size and production demands on go-ipfs
+   without _any_ large-scale testing or network simulation infrastructure. In
+   the past all production scale testing has been done by deploying a custom
+   go-ipfs build to a bootstrapper or gateway and watching how it behaves.
+   
+To address these concerns, we've put a halt to all non-bugfix go-ipfs releases
+as we improve our testing practices and build out our testing and network
+simulation infrastructure. We felt this measure necessary as testing tends to
+become a secondary concern in the face new features and pressing performance
+issues.
+
+However, even with our current testing practices, these regressions should have
+been caught by pre-release testing. Furthermore, a better release process (e.g.,
+the ability to cut patch releases) would have enabled us to quickly release
+fixes for these regressions.
```

Comment:
This is also about communicating and managing expectations that the communty can have about a release, without expecting them to have read a blog post about it. With semver the community can (relatively) safely assume that they can upgrade to a patch version without it breaking any of their code.

It gives us the flexibility to release patches without the fanfare of a major release and will probably encourage more users to upgrade sooner knowing that upgrading will be easy.

Reply:
LGTY?